### PR TITLE
Added docker configuration files for containerisation of the applicat…

### PR DESCRIPTION
**Pull request by kogoe1**
…ion. Also modified app.py to make it run in gunicorn

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+#docker
+Dockerfile
+.dockerignore
+
+#git
+.git
+.gitignore
+
+
```
Comment (hugh-emerson) [inline .dockerignore:29]: You should add .venv to avoid copying your local virtual environment into the image

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+RUN apt-get update && apt-get install -y curl
+RUN curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python
+EXPOSE 5000
+ENV PATH="${PATH}:/root/.poetry/bin"
+RUN mkdir -p /devops_mod5
+WORKDIR /devops_mod5
+COPY . /devops_mod5
+RUN poetry install 
+ENV VIRTUAL_ENV=.venv
+ENV PATH="$VIRTUAL_ENV/bin:$PATH"
```
Comment (hugh-emerson) [inline Dockerfile:11]: You should be able to remove these lines if you run gunicorn with `poetry run gunicorn`.

But better yet, you don't need a virtual environment inside your Docker container because it's already isolated. So you could instead use this for your production stage 
```
RUN poetry config virtualenvs.create false --local && poetry install --no-dev --no-root
```
Note the `--no-dev` flag is to skip packages like `pytest` assuming the pyproject.toml file is set up correctly.

File: pyproject.toml [hunk 3]
```diff
@@ -10,6 +10,8 @@ flask = "^1.1.2"
 python-dotenv = "^0.14.0"
 Flask-WTF = "^0.14.3"
 requests = "^2.25.0"
+gunicorn = "^20.0.4"
+pytest = "^6.2.3"
```
Comment (hugh-emerson) [inline pyproject.toml:14]: pytest is already included in the dev-dependencies section below

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+FROM base AS production
+ENTRYPOINT ["gunicorn", "-w", "3", "-b", "0.0.0.0:5000", "todo_app.app:app"]
```
Comment (hugh-emerson) [inline Dockerfile:14]: Why have you got rid of your create_app function in app.py? You can use a function in your gunicorn command:

```suggestion
ENTRYPOINT ["gunicorn", "-w", "3", "-b", "0.0.0.0:5000", "todo_app.app:create_app()"]
```

File: todo_app/app.py [hunk 5]
```diff
@@ -6,71 +6,71 @@
...
 from todo_app.data.model import ViewModel
 
+from dotenv import load_dotenv, find_dotenv
+
 import os
 
 
+app = Flask(__name__)
 
-def create_app():
```
Comment (hugh-emerson) [inline todo_app/app.py:13]: Moving everything out of `create_app` will break your tests. Undo this change

**Review by hugh-emerson — APPROVED** [review]
You really need to put `create_app` back in for your tests to work. See my comment on the Dockerfile for how to get that working with gunicorn.

Otherwise I've just left some minor suggestions, so once you fix that, you can go ahead and merge 👍

File: .dockerignore (hunk 1 above)
Comment (kogoe1) [reply .dockerignore:29]: This is now added

File: Dockerfile (hunk 2 above)
Comment (kogoe1) [reply Dockerfile:11]: Thanks Hugh
Have removed the virtual env ENV config and added the suggested RUN line

File: pyproject.toml (hunk 3 above)
Comment (kogoe1) [reply pyproject.toml:14]: That purest line is now removed

File: Dockerfile (hunk 4 above)
Comment (kogoe1) [reply Dockerfile:14]: That was the first thing I tried but it wouldn't work. I wasted a lot of time trying to get it work so I removed it.
Just realised all I needed to do was to add "()" to `create_app` when running the gunicorn command

Thanks a lot for this Hugh! Won't need to rewrite my tests now :)

File: todo_app/app.py (hunk 5 above)
Comment (kogoe1) [reply todo_app/app.py:13]: This has been added back in

File: todo_app/app.py [hunk 6]
```diff
@@ -6,31 +6,34 @@
...
-
+from dotenv import load_dotenv, find_dotenv
 
+import os
 
 def create_app():
     app = Flask(__name__)
 
+    file_path = find_dotenv('.env')
+    load_dotenv(file_path, override=True)
```
Comment (Jonesey13) [inline todo_app/app.py:17]: These lines currently break the selenium tests (they should really be in the `test_app()` method)

**Review by Jonesey13 — COMMENTED** [review]
Looks good to me barring an issue with the selenium test (see below). I'll pass this submission but please fix the test issue.

File: todo_app/app.py (hunk 6 above)
Comment (kogoe1) [reply todo_app/app.py:17]: Thanks for pointing that out Alex. Have removed those lines from `app.py`. These lines are already in `test_selenium.py` (lines 19 and 20)
 Made this change via https://github.com/kogoe1/DevOps-Course-Starter/pull/7 and have merged into `master`
